src/main/java/net/consensys/orion/http/handler/privacy/PrivacyGroupHandler.java

```diff
@@ -14,56 +14,150 @@
 
+
+  private static final Logger log = LogManager.getLogger();
+
   private final Storage<PrivacyGroupPayload> privacyGroupStorage;
+  private final ConcurrentNetworkNodes networkNodes;
+  private final Enclave enclave;
+  private final HttpClient httpClient;
 
-  public PrivacyGroupHandler(Storage<PrivacyGroupPayload> privacyGroupStorage) {
+  public PrivacyGroupHandler(
+      Storage<PrivacyGroupPayload> privacyGroupStorage,
+      ConcurrentNetworkNodes networkNodes,
+      Enclave enclave,
+      Vertx vertx,
+      Config config) {
     this.privacyGroupStorage = privacyGroupStorage;
+    this.networkNodes = networkNodes;
+    this.enclave = enclave;
+    this.httpClient = NodeHttpClientBuilder.build(vertx, config, 1500);
   }
 
   @Override
   public void handle(RoutingContext routingContext) {
 
     byte[] request = routingContext.getBody().getBytes();
-    PrivacyGroupRequest addressList = Serializer.deserialize(JSON, PrivacyGroupRequest.class, request);
+    PrivacyGroupRequest privacyGroup = Serializer.deserialize(JSON, PrivacyGroupRequest.class, request);
 
-    SecureRandom random = new SecureRandom();
-    byte bytes[] = new byte[20];
-    random.nextBytes(bytes);
+    byte[] bytes = new byte[20];
+    if (privacyGroup.getSeed().isPresent()) {
+      bytes = privacyGroup.getSeed().get();
+    } else {
+      SecureRandom random = new SecureRandom();
+      random.nextBytes(bytes);
+    }
 
     PrivacyGroupPayload privacyGroupPayload = new PrivacyGroupPayload(
-        addressList.addresses(),
+        privacyGroup.addresses(),
+        privacyGroup.name(),
+        privacyGroup.description(),
         PrivacyGroupPayload.State.ACTIVE,
         PrivacyGroupPayload.Type.PANTHEON,
         bytes);
     final String privacyGroupId = privacyGroupStorage.generateDigest(privacyGroupPayload);
 
-    privacyGroupStorage.put(privacyGroupPayload).thenAccept((result) -> {
+    List<Box.PublicKey> addressListToForward = Arrays
+        .stream(privacyGroup.addresses())
+        .filter(key -> !key.equals(privacyGroup.from())) // don't forward to self
+        .distinct()
+        .map(enclave::readKey)
+        .collect(Collectors.toList());
+
+    if (addressListToForward.stream().anyMatch(pKey -> networkNodes.urlForRecipient(pKey) == null)) {
+      routingContext.fail(new OrionException(OrionErrorCode.NODE_MISSING_PEER_URL, "couldn't find peer URL "));
+      return;
+    }
+    // propagate payload
+    log.debug("propagating payload");
+
+    @SuppressWarnings("rawtypes")
+    CompletableFuture[] cfs = addressListToForward.stream().map(pKey -> {
+      URL recipientURL = networkNodes.urlForRecipient(pKey);
+
+      CompletableFuture<Boolean> responseFuture = new CompletableFuture<>();
+
+      // serialize payload, stripping non-relevant encryptedKeys, and configureRoutes payload
+      final byte[] payload = Serializer.serialize(HttpContentType.CBOR, privacyGroupPayload);
+
+      // execute request
+      httpClient
```

Comment:
If a node fails why processing this message, or receiving it on the other end, can the 2 nodes end up in different states?  If so, Is there a mechanism to sync things up later?

Reply:
If the node fails on receiving end, it will not be updated even in the current storage and will exit with error. We should invoke the API again to create the group.